qiskit/quantum_info/operators/symplectic/clifford.py

```diff
@@ -848,6 +863,72 @@ def _from_label(label):
         symp[-1] = phase
         return symp
 
+    @staticmethod
+    def _unitary_matrix_to_tableau(matrix):
+        # pylint: disable=invalid-name
+        U = Operator(matrix)
+        Udg = U.adjoint()
+        num_qubits = U.num_qubits
+        # Create full table for a n-qubit Pauli matrix to a row of the tableau
+        x_table = {
+            "I": False,
+            "X": True,
+            "Z": False,
+            "Y": True,
+        }
+        z_table = {
+            "I": False,
+            "X": False,
+            "Z": True,
+            "Y": True,
+        }
+        matrix_row_pairs = []
+        for paulis in itertools.product(["I", "X", "Y", "Z"], repeat=num_qubits):
+            # positive phase Pauli case
+            row_p = np.empty(2 * num_qubits + 1, dtype=bool)
+            phase = False
+            for i, pauli in enumerate(paulis):
+                row_p[i] = x_table[pauli]
+                row_p[i + num_qubits] = z_table[pauli]
+                phase ^= pauli == "Y"
+            row_p[-1] = phase
+            pauli_str = "".join(reversed(paulis))
+            pauli_op = Operator.from_label(pauli_str)
+            matrix_p = ((-1) ** phase) * pauli_op.to_matrix()
+            matrix_row_pairs.append((matrix_p, row_p))
+            # negative phase Pauli case
+            row_n = np.copy(row_p)
+            row_n[-1] ^= True
+            matrix_row_pairs.append(((-1) * matrix_p, row_n))
+
+        # Construct stabilizer/destabilizer tables
+        stab = np.empty((num_qubits, 2 * num_qubits + 1), dtype=bool)
+        for i in range(num_qubits):
+            label = "I" * (num_qubits - i - 1) + "X" + "I" * i
+            Xi = Operator.from_label(label).to_matrix()
+            target = U @ Xi @ Udg
+            for mat, row in matrix_row_pairs:
+                if np.allclose(mat, target):
+                    stab[i] = row
+                    break
+            else:
+                raise QiskitError("Non-Clifford matrix is not convertible")
+
```

Comment:
One more quick question. I have not done the math, but I am wondering if it's possible to check whether a given matrix (here `U * Xi * Udg`) is close to some Pauli matrix without explicitly going over all of the Pauli matrices?

Reply:
These lines are using a bit strange grammar of Python: This `else` clause is coupled with `for` not `if`. So they do go over all of the Pauli matrices.

Reply:
Sorry, I did not highlight the right lines. I believe (but I have not done the math) that the (tensor products of) Pauli matrices are very special unitary matrices, and it should be possible to check whether a given matrix (either `U * Xi * Udg` or `U * Zi * Udg`) is close to a Pauli matrix without explicitly checking against the 2*4^n matrices.

Reply:
Yeah, I'm a bit concerned that there may be a more efficient algorithm then current one. If anyone knows such an algorithm, please let me know.

Reply:
I believe that if you look at the 2^N x 2^N unitary matrix corresponding to a Pauli matrix, then this unitary matrix must have only one nonzero entry in every row and every column, moreover each nonzero element can only be 1, -1, i or -i. Intuitively, I think this should be enough to work out the exact Pauli.

Reply:
Great suggestion, thanks! The new algorithm should be 
1. Check if the `U * Xi * Udg` (or `U * Zi * Udg`) has exactly 2^N non-zero (i.e. 1, -1, i or -i) elements (if not, U is non-Clifford)
3. Reconstruct the Pauli string (a row of the resulting tableau) from the positions and values of the 2^N non-zero elements (if impossible, U is non-Clifford)
I'll try it.

Reply:
Thank you! Here are my five cents (and @ikkoham and @ShellyGarion are welcome to correct this): 
* If a gate is going to be used a lot, it might be best to provide methods that directly operate on the Clifford tableau (like what is done in `_append_cx` or `_append_w` )
* However, having a generic fallback mechanism is a great idea, so I think this development is very useful
* If the gate has a `definition`, then it might be more efficient to consider the `definition` first, before constructing the unitary matrix and converting it to Clifford. On the other hand, it is possible that the default definition of a Clifford gate might contain non-Clifford gates, so the construction based on the definition alone might not be possible

Reply:
Regarding the more efficient O(4^n) (instead of O(16^n)) algorithm, implemented at [bad10de](https://github.com/Qiskit/qiskit-terra/pull/9475/commits/bad10dec64c2d6a72f27f3a6125bc00e15ca362d)